File: skillswap-client/src/services/userService.ts
import api from './api';

export interface UserProfile {
    id: string;
    email: string;
    fullName: string;
    bio?: string;
    timeBalance: number;
    rating: number;
    skills: SkillSummary[];
}

export interface SkillSummary {
    id: number;
    title: string;
    isOffering: boolean;
    categoryName: string;
}

export interface UserSearchResult {
    id: string;
    fullName: string;
    bio?: string;
    rating: number;
    skills: SkillSummary[];
}

export interface UpdateProfileDto {
    fullName?: string;
    bio?: string;
}

export interface UserSearchParams {
    name?: string;
    skill?: string;
    page?: number;
    pageSize?: number;
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
}

export const userService = {
    getMe: () =>
        api.get<UserProfile>('/users/me'),
    getProfile: (id: string) =>
        api.get<UserProfile>(`/users/${id}`),
    updateProfile: (dto: UpdateProfileDto) =>
        api.put<UserProfile>('/users/profile', dto),
    searchUsers: (params?: UserSearchParams) =>
        api.get<PagedResult<UserSearchResult>>('/users/search', { params }),
};

File: SkillSwap.API/Services/UserService.cs
using SkillSwap.Core.DTOs.Common;
using SkillSwap.Core.DTOs.Users;
using SkillSwap.Core.Exceptions;
using SkillSwap.Core.Interfaces.Repositories;
using SkillSwap.Core.Interfaces.Services;

namespace SkillSwap.API.Services;

public class UserService : IUserService
{
    private readonly IUserRepository _userRepo;

    public UserService(IUserRepository userRepo) => _userRepo = userRepo;

    public async Task<UserProfileDto> GetProfileAsync(string userId)
    {
        var user = await _userRepo.GetByIdWithSkillsAsync(userId)
            ?? throw new NotFoundException("User", userId);

        return new UserProfileDto
        {
            Id = user.Id,
            Email = user.Email!,
            FullName = user.FullName,
            Bio = user.Bio,
            TimeBalance = user.TimeBalance,
            Rating = user.Rating,
            Skills = user.Skills.Select(s => new SkillSummaryDto
            {
                Id = s.Id,
                Title = s.Title,
                IsOffering = s.IsOffering,
                CategoryName = s.Category?.Name ?? ""
            }).ToList()
        };
    }

    public async Task<UserProfileDto> UpdateProfileAsync(string userId, UpdateUserProfileDto dto)
    {
        var user = await _userRepo.GetByIdAsync(userId)
            ?? throw new NotFoundException("User", userId);

        if (dto.FullName != null) user.FullName = dto.FullName;
        if (dto.Bio != null) user.Bio = dto.Bio;

        await _userRepo.UpdateAsync(user);
        return await GetProfileAsync(userId);
    }

    public async Task<PagedResult<UserSearchResultDto>> SearchUsersAsync(UserSearchParams searchParams)
    {
        var (users, totalCount) = await _userRepo.SearchUsersAsync(
            searchParams.Name, searchParams.Skill,
            searchParams.Page, searchParams.PageSize);

        var items = users.Select(u => new UserSearchResultDto
        {
            Id = u.Id,
            FullName = u.FullName,
            Bio = u.Bio,
            Rating = u.Rating,
            Skills = u.Skills.Select(s => new SkillSummaryDto
            {
                Id = s.Id,
                Title = s.Title,
                IsOffering = s.IsOffering,
                CategoryName = s.Category?.Name ?? ""
            }).ToList()
        });
        
        return new PagedResult<UserSearchResultDto>
        {
            Items = items,
            TotalCount = totalCount
        };
    }
}
